### PR TITLE
Support `Multilinestring` in column API and python Bindings for `pairwise_linestring_distance`

### LINKED ISSUE #706: Simplify header-only linestring_distance API, support multilinestring
Header only Point distance API should be updated to support multilinestring to multilinestring distance, conform to geoarrow format and adopt iterator collection structure to simplify the interface.

### PR DESCRIPTION
**Pull request by isVoid**
## Description

This is a port of the second half of #753, closes #706

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by isVoid** [comment]
Note that `linestring_distance_test_medium.cu` has over 1300+ loc. If you mask out this file actual line changes is not that big.

File: cpp/include/cuspatial/distance/linestring_distance.hpp [hunk 1]
```diff
@@ -57,65 +56,67 @@ namespace cuspatial {
...
  *
  */
 std::unique_ptr<cudf::column> pairwise_linestring_distance(
-  cudf::device_span<cudf::size_type const> linestring1_offsets,
-  cudf::column_view const& linestring1_points_x,
-  cudf::column_view const& linestring1_points_y,
-  cudf::device_span<cudf::size_type const> linestring2_offsets,
-  cudf::column_view const& linestring2_points_x,
-  cudf::column_view const& linestring2_points_y,
+  std::optional<cudf::device_span<cudf::size_type const>> multilinestring1_geometry_offsets,
```
Comment (harrism) [inline cpp/include/cuspatial/distance/linestring_distance.hpp:115]: Perhaps mention in docstring why these are optional?

**Review by harrism — APPROVED** [review]
Looks great!

**Review by thomcom — APPROVED** [review]

**Comment by isVoid** [comment]
@gpucibot merge

**Pull request merged**